src/Common/src/System/Text/OSEncoder.cs

```diff
@@ -0,0 +1,262 @@
+            {
+                if (count <= 0)
+                    return 0;
+
+                return OSEncoding.WideCharToMultiByte(_encoding.CodePage, chars, count, null, 0);
+            }
+
+            // we have left over character
+            if (count == 0 && !excludeLastChar && !flush)
+                return 0;
+
+            char [] bufferToProcess = new char[count+1];
+            bufferToProcess[0] = _charLeftOver;
+            fixed (char *pBuffer = bufferToProcess)
+            {
+                Buffer.MemoryCopy(chars, pBuffer+1, count * sizeof(char), count * sizeof(char));
+                return OSEncoding.WideCharToMultiByte(_encoding.CodePage, pBuffer, bufferToProcess.Length, null, 0);
+            }
+        }
+
+        public override unsafe int GetBytes(char[] chars, int charIndex, int charCount,
+                                              byte[] bytes, int byteIndex, bool flush)
+        {
+            if (chars == null || bytes == null)
+                throw new ArgumentNullException(chars == null ? nameof(chars) : nameof(bytes), SR.ArgumentNull_Array);
+
+            if (charIndex < 0 || charCount < 0)
+                throw new ArgumentOutOfRangeException(charIndex < 0 ? nameof(charIndex) : nameof(charCount), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            if (chars.Length - charIndex < charCount)
+                throw new ArgumentOutOfRangeException(nameof(chars), SR.ArgumentOutOfRange_IndexCountBuffer);
+
+            if (byteIndex < 0 || byteIndex > bytes.Length)
+                throw new ArgumentOutOfRangeException(nameof(byteIndex), SR.ArgumentOutOfRange_Index);
+            
+            if (bytes.Length == 0)
+                return 0;
+
+            if (charCount == 0 && (_charLeftOver == NULL_CHAR || !flush))
+                return 0;
+            
+            fixed (char* pChars = chars)
+            fixed (byte* pBytes = bytes)
+            {
+                char *pSingleElementArray = stackalloc char[1]; // to avoid passing null to GetByteCount
+                char *pBuffer = pChars == null ?  pSingleElementArray : pChars + charIndex;
+
+                return GetBytes(pBuffer, charCount, pBytes + byteIndex, bytes.Length - byteIndex, flush);
+            }
+        }
+
+        public unsafe override int GetBytes(char* chars, int charCount, byte* bytes, int byteCount, bool flush)
+        {
+            if (chars == null || bytes == null)
+                throw new ArgumentNullException(chars == null ? nameof(chars) : nameof(bytes), SR.ArgumentNull_Array);
+
+            if (byteCount < 0 || charCount < 0)
+                throw new ArgumentOutOfRangeException(byteCount < 0 ? nameof(byteCount) : nameof(charCount), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            if (byteCount == 0)
+                return 0;
+
+            char lastChar = charCount > 0 && !flush && Char.IsHighSurrogate(chars[charCount - 1]) ? chars[charCount - 1] : NULL_CHAR;
```

Comment:
yes it should work correctly even the source array contains nulls. lastChar is kind of flag either has upper surrogate or null. if it is null, we don't use it at all in the encoding conversion.

---
In reply to: [93727921](https://github.com/dotnet/corefx/pull/14649#discussion_r93727921) [](ancestors = 93727921)